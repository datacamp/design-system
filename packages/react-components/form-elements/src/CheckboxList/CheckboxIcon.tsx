import { CheckmarkIcon } from '@datacamp/waffles-icons';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

const CheckboxIcon: React.FC<{
  checked: boolean;
  disabled: boolean;
  error: boolean;
}> = ({ checked, disabled, error }) => {
  let color = tokens.color.opaque.greyLight.value.hex;
  if (checked) color = tokens.color.opaque.primary.value.hex;
  if (error) color = tokens.color.opaque.red.value.hex;
  if (disabled) color = tokens.color.opaque.greyLighter.value.hex;

  return (
    <>
      <span
        css={css({
          // some of these styles will be overriden when focussed etc.
          alignItems: 'center',
          borderColor: color,
          borderRadius: '4px',
          borderStyle: 'solid',
          borderWidth: '1px',
          boxSizing: 'border-box',
          color,
          display: 'flex',
          height: '16px',
          justifyContent: 'center',
          left: 1,
          position: 'absolute',
          top: 1,
          width: '16px',
        })}
      >
        {checked && (
          <CheckmarkIcon
            aria-hidden={true} // Hide this icon, as label and input element are sufficient
            size={12}
            title="" // Remove the title, as label and input element are sufficient
          />
        )}
      </span>
    </>
  );
};

export default CheckboxIcon;
