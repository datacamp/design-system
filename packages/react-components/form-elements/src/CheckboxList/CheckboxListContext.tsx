import React from 'react';

export interface CheckboxListContextInterface {
  disabled: boolean;
  hasError: boolean;
  name: string;
  onChange: (value: string) => void;
  value: string[];
}

const CheckboxListContext = React.createContext<CheckboxListContextInterface | null>(
  null,
);

export default CheckboxListContext;
