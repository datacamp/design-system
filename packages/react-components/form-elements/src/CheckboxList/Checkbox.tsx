import { Text } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { ssrSafeNotFirstChildSelector } from '@datacamp/waffles-utils';
import { css } from '@emotion/core';
import _ from 'lodash';
import React, { ReactElement } from 'react';

import CheckboxIcon from './CheckboxIcon';
import CheckboxListContext from './CheckboxListContext';

export interface CheckboxProps {
  /**
   * The text to display for this item.
   */
  children: string;
  /**
   * When true this individual option will be disabled. This is overriden when
   * disabled is set on CheckboxList itself.
   */
  disabled?: boolean;
  /**
   * The value of this option. If this matches the value provided to the parent
   * `CheckboxList`, this option will be selected.
   */
  value: string;
}

const textStyle = css({
  marginLeft: tokens.size.space[24].value,
});

const divStyle = css({
  alignItems: 'center',
  display: 'flex',
  height: '20px',
  marginTop: 0,
  position: 'relative',
  [ssrSafeNotFirstChildSelector]: { marginTop: '10px' },
});

/**
 * Can only be used within `CheckboxList`
 */
const Checkbox = ({
  children,
  disabled = false,
  value,
}: CheckboxProps): ReactElement => {
  const contextValue = React.useContext(CheckboxListContext);
  if (contextValue === null) {
    throw new Error('Checkbox must be used with CheckboxList');
  }

  const elementDisabled = disabled || contextValue.disabled;
  const handleChange = (): void => contextValue.onChange(value);
  const isChecked = _.includes(contextValue.value, value);

  const focusColor = contextValue.hasError
    ? tokens.color.primary.redDark.value.hex
    : tokens.color.primary.blueDark.value.hex;

  const focusStyle = {
    borderColor: focusColor,
    color: focusColor,
  };

  return (
    <label css={divStyle} htmlFor={value}>
      <input
        checked={isChecked}
        css={css(
          {
            opacity: 0,
            position: 'absolute',
            width: 0,
          },
          !disabled && {
            // affects the style of the CheckboxIcon
            ':active + span': focusStyle,
            ':focus + span': focusStyle,
          },
        )}
        disabled={elementDisabled}
        id={value}
        name={contextValue.name}
        onChange={handleChange}
        type="checkbox"
      />
      <CheckboxIcon
        checked={isChecked}
        disabled={disabled || contextValue.disabled}
        error={contextValue.hasError}
      />
      <Text
        css={css(
          textStyle,
          elementDisabled && {
            opacity: 0.3,
          },
        )}
      >
        {children}
      </Text>
    </label>
  );
};

export default Checkbox;
