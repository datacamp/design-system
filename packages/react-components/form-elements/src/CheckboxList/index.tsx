import { childrenOfType } from 'airbnb-prop-types';
import _ from 'lodash';
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

import Label from '../Label';

import Checkbox, { CheckboxProps } from './Checkbox';
import CheckboxListContext from './CheckboxListContext';

export interface CheckboxListProps {
  children: Array<ReactElement<CheckboxProps>>;
  /**
   * When provided the description will show underneath the label. Only
   * available when 'label' is also specified.
   */
  description?: string;
  /**
   * Disables the whole CheckboxList when true.
   */
  disabled?: boolean;
  /**
   * When provided the errorMessage will show underneath the CheckboxList.
   */
  errorMessage?: string;
  /**
   * Renders label text above the CheckboxList.
   */
  label: string;
  /**
   * Sets the name attribute on the rendered Checkbox buttons.
   */
  name: string;
  /**
   * Called when the user requests a change to the value of the CheckboxList. This
   * should be used to set the value.
   */
  onChange: (value: string[]) => void;
  /**
   * It defines whether the CheckboxList is required or not. If required=true it
   * adds the text 'Required' on the top-right of the CheckboxList, if
   * required=false it adds 'Optional'. The default value is undefined, which
   * doesn't add anything.
   */
  required?: boolean;
  /**
   * The current value of the CheckboxList. This should be controlled by listening to
   * onChange.
   */
  value: string[];
}

const CheckboxList = ({
  children,
  description,
  disabled = false,
  errorMessage,
  label,
  name,
  onChange,
  required,
  value,
}: CheckboxListProps): ReactElement => {
  const handleChange = React.useCallback(
    (changedValue: string): void => {
      // removes if already exists, adds if it doesn't
      const newValue = _.includes(value, changedValue)
        ? _.filter(value, (item) => item !== changedValue)
        : _.union(value, [changedValue]);

      onChange(newValue);
    },
    [onChange, value],
  );

  return (
    <Label
      as="fieldset"
      description={description}
      errorMessage={errorMessage}
      label={label}
      required={required}
    >
      <div>
        <CheckboxListContext.Provider
          value={{
            disabled,
            hasError: !!errorMessage,
            name,
            onChange: handleChange,
            value,
          }}
        >
          {children}
        </CheckboxListContext.Provider>
      </div>
    </Label>
  );
};

CheckboxList.propTypes = {
  /**
   * Takes multiple `Checkbox` components as children.
   */
  children: PropTypes.arrayOf(childrenOfType(Checkbox).isRequired).isRequired,
};

export default CheckboxList;
