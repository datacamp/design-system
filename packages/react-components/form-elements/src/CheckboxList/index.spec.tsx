import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import Checkbox from './Checkbox';
import CheckboxList from './index';

const testName = 'test-name';
const testValue = ['test value'];
const testLabel = 'test label';

describe('<CheckboxList>', () => {
  it('renders the list', () => {
    const { getByLabelText, getByText, debug } = render(
      <CheckboxList
        label={testLabel}
        name={testName}
        onChange={() => {}}
        value={testValue}
      >
        <Checkbox value={testValue[0]}>item 1</Checkbox>
        <Checkbox value="value2">item 2</Checkbox>
      </CheckboxList>,
    );

    expect(getByText(testLabel) as HTMLElement).toBeInTheDocument();
    debug();
    const item1 = getByLabelText('item 1') as HTMLElement;
    const item2 = getByLabelText('item 2') as HTMLElement;

    // all elements have the name
    expect(item1).toHaveAttribute('name', testName);
    expect(item2).toHaveAttribute('name', testName);
  });

  describe('required', () => {
    it('renders the indication "Required" above the checkbox group if required is passed as a prop', () => {
      const { getByText } = render(
        <CheckboxList
          label={testLabel}
          name={testName}
          onChange={() => {}}
          required
          value={testValue}
        >
          <Checkbox value={testValue[0]}>item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
        </CheckboxList>,
      );

      expect(getByText('Required')).toBeInTheDocument();
    });
    it('renders the indication "Optional" above the checkbox group if required=false is passed as a prop', () => {
      const { getByText } = render(
        <CheckboxList
          label={testLabel}
          name={testName}
          onChange={() => {}}
          required={false}
          value={testValue}
        >
          <Checkbox value={testValue[0]}>item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
        </CheckboxList>,
      );

      expect(getByText('Optional')).toBeInTheDocument();
    });
  });

  it('adds an error message under the checkbox group if error is passed as a prop', () => {
    const testError = 'Test error';

    const { getByText } = render(
      <CheckboxList
        errorMessage={testError}
        label={testLabel}
        name={testName}
        onChange={() => {}}
        value={testValue}
      >
        <Checkbox value={testValue[0]}>item 1</Checkbox>
        <Checkbox value="value2">item 2</Checkbox>
      </CheckboxList>,
    );
    expect(getByText(testError)).toBeInTheDocument();
  });

  it('disables the whole component', () => {
    const { getByLabelText } = render(
      <CheckboxList
        disabled
        label="test-label"
        name={testName}
        onChange={() => {}}
        value={['value1']}
      >
        <Checkbox value="value1">item 1</Checkbox>
        <Checkbox value="value2">item 2</Checkbox>
      </CheckboxList>,
    );

    const radioElement1 = getByLabelText('item 1') as HTMLElement;
    const radioElement2 = getByLabelText('item 2') as HTMLElement;

    expect(radioElement1).toHaveAttribute('disabled');
    expect(radioElement2).toHaveAttribute('disabled');
  });

  it('disables an individual option', () => {
    const { getByLabelText } = render(
      <CheckboxList
        label="test-label"
        name={testName}
        onChange={() => {}}
        value={['value1']}
      >
        <Checkbox value="value1">item 1</Checkbox>
        <Checkbox disabled value="value2">
          item 2
        </Checkbox>
      </CheckboxList>,
    );

    const radioElement1 = getByLabelText('item 1') as HTMLElement;
    const radioElement2 = getByLabelText('item 2') as HTMLElement;

    expect(radioElement1).not.toHaveAttribute('disabled');
    expect(radioElement2).toHaveAttribute('disabled');
  });

  describe('interactivity', () => {
    it('has no selected option when value is invalid', () => {
      const { getByLabelText } = render(
        <CheckboxList
          label="test-label"
          name={testName}
          onChange={() => {}}
          value={['']}
        >
          <Checkbox value="value1">item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
        </CheckboxList>,
      );

      const radioElement1 = getByLabelText('item 1') as HTMLElement;
      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      expect(radioElement1).not.toHaveAttribute('checked');
      expect(radioElement2).not.toHaveAttribute('checked');
    });

    it('selects the correct option for the value', () => {
      const { getByLabelText } = render(
        <CheckboxList
          label="test-label"
          name={testName}
          onChange={() => {}}
          value={['value1']}
        >
          <Checkbox value="value1">item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
        </CheckboxList>,
      );

      const radioElement1 = getByLabelText('item 1') as HTMLElement;
      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      expect(radioElement1).toHaveAttribute('checked');
      expect(radioElement2).not.toHaveAttribute('checked');
    });

    it('calls onChange with the selected value', () => {
      const onChange = jest.fn();
      const { getByLabelText } = render(
        <CheckboxList
          label="test-label"
          name={testName}
          onChange={onChange}
          value={['value1']}
        >
          <Checkbox value="value1">item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
        </CheckboxList>,
      );

      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      userEvent.click(radioElement2);

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith(['value1', 'value2']);
    });
  });

  describe('snapshots', () => {
    it(`renders a checkbox list`, async () => {
      const { container } = render(
        <CheckboxList
          errorMessage="test error"
          label={testLabel}
          name={testName}
          onChange={() => {}}
          value={testValue}
        >
          <Checkbox value={testValue[0]}>item 1</Checkbox>
          <Checkbox value="value2">item 2</Checkbox>
          <Checkbox disabled value="value2">
            item 3
          </Checkbox>
        </CheckboxList>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
