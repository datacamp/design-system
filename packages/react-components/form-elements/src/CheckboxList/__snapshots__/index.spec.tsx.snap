// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxList> snapshots renders a checkbox list 1`] = `
.emotion-16 {
  border: 0;
  display: block;
  margin: 0;
  margin-top: 0;
  min-width: 0;
  padding: 0;
}

*:not(style) ~ .emotion-16 {
  margin-top: 16px;
}

.emotion-2 {
  display: block;
  margin-bottom: 12px;
  padding: 0;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  color: #3d4251;
  display: inline-block;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  font-weight: 700;
  margin-right: 4px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-6 {
  color: #d1d3d8;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 0;
  position: relative;
}

*:not(style) ~ .emotion-6 {
  margin-top: 14px;
}

.emotion-3 {
  opacity: 0;
  position: absolute;
  width: 0;
}

.emotion-3:active + span {
  border-width: 2px;
  height: 18px;
  left: 0;
  top: 0;
  width: 18px;
}

.emotion-3:focus + span {
  border-width: 2px;
  height: 18px;
  left: 0;
  top: 0;
  width: 18px;
}

.emotion-5 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  margin-left: 24px;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: #fe5c5c;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-sizing: border-box;
  color: #fe5c5c;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 16px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 1px;
  position: absolute;
  top: 1px;
  width: 16px;
}

.emotion-15 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  color: #fe5c5c;
  display: block;
  font-size: 14px;
  margin-top: 8px;
}

.emotion-12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: #e6eaeb;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-sizing: border-box;
  color: #e6eaeb;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 16px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 1px;
  position: absolute;
  top: 1px;
  width: 16px;
}

.emotion-13 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  margin-left: 24px;
  color: #d1d3d8;
}

.emotion-11 {
  opacity: 0;
  position: absolute;
  width: 0;
}

<fieldset
  class="emotion-16"
>
  <legend
    class="emotion-2"
  >
    <span
      class="emotion-1"
    >
      <span
        class="emotion-0"
      >
        test label
      </span>
    </span>
  </legend>
  <div>
    <label
      class="emotion-6"
      for="test value"
    >
      <input
        checked=""
        class="emotion-3"
        id="test value"
        name="test-name"
        type="checkbox"
      />
      <span
        class="emotion-4"
      >
        <svg
          aria-hidden="true"
          height="12"
          role="img"
          viewBox="0 0 18 18"
          width="12"
        >
          <path
            d="M13.746 4.337a1.015 1.015 0 011.409-.099c.417.354.462.97.101 1.378l-7.13 8.047a1.015 1.015 0 01-1.483.03L2.771 9.67a.961.961 0 01.044-1.38 1.015 1.015 0 011.412.041l3.113 3.235 6.406-7.229z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="emotion-5"
      >
        item 1
      </span>
    </label>
    <label
      class="emotion-6"
      for="value2"
    >
      <input
        class="emotion-3"
        id="value2"
        name="test-name"
        type="checkbox"
      />
      <span
        class="emotion-4"
      />
      <span
        class="emotion-5"
      >
        item 2
      </span>
    </label>
    <label
      class="emotion-6"
      for="value2"
    >
      <input
        class="emotion-11"
        disabled=""
        id="value2"
        name="test-name"
        type="checkbox"
      />
      <span
        class="emotion-12"
      />
      <span
        class="emotion-13"
      >
        item 3
      </span>
    </label>
  </div>
  <span
    class="emotion-15"
  >
    test error
  </span>
</fieldset>
`;
